File: lib/share.ts
import fs from 'fs/promises';
import path from 'path';
import crypto from 'crypto';

const SHARE_LINKS_FILE = path.join(process.cwd(), 'share-links.json');

let _lock: Promise<void> = Promise.resolve();
function withLock<T>(fn: () => Promise<T>): Promise<T> {
  const prev = _lock;
  let release: () => void;
  _lock = new Promise<void>((r) => { release = r; });
  return prev.then(fn).finally(() => release!());
}

export interface ShareLink {
  path: string;
  type: 'article' | 'folder';
  createdAt: string;
  /** PIN 码 sha256 hex（6-20 位明文 PIN 的哈希）。为空表示无密码 */
  pinHash?: string;
  /** ISO 时间戳；缺失视为永久 */
  expiresAt?: string;
}

export interface CreateShareOptions {
  pin?: string;              // 明文 PIN
  expiresInDays?: number;    // 7 / 30 / null=永久
}

async function initShareLinksFile(): Promise<void> {
  try {
    await fs.stat(SHARE_LINKS_FILE);
  } catch {
    await fs.writeFile(SHARE_LINKS_FILE, JSON.stringify({}), 'utf-8');
  }
}

async function getAllShareLinks(): Promise<Record<string, ShareLink>> {
  await initShareLinksFile();
  const content = await fs.readFile(SHARE_LINKS_FILE, 'utf-8');
  return JSON.parse(content || '{}');
}

async function saveShareLinks(links: Record<string, ShareLink>): Promise<void> {
  await fs.writeFile(SHARE_LINKS_FILE, JSON.stringify(links, null, 2), 'utf-8');
}

function hashPin(pin: string): string {
  return crypto.createHash('sha256').update(`nexo:${pin}`).digest('hex');
}

export async function createShareLink(
  itemPath: string,
  type: 'article' | 'folder',
  opts: CreateShareOptions = {},
): Promise<string> {
  return withLock(async () => {
    const links = await getAllShareLinks();
    let token: string;
    do {
      token = crypto.randomBytes(16).toString('hex');
    } while (links[token]);

    const entry: ShareLink = {
      path: itemPath,
      type,
      createdAt: new Date().toISOString(),
    };
    if (opts.pin && opts.pin.trim()) entry.pinHash = hashPin(opts.pin.trim());
    if (opts.expiresInDays && opts.expiresInDays > 0) {
      entry.expiresAt = new Date(Date.now() + opts.expiresInDays * 86400000).toISOString();
    }

    links[token] = entry;
    await saveShareLinks(links);
    return token;
  });
}

export async function deleteShareLink(token: string): Promise<boolean> {
  return withLock(async () => {
    const links = await getAllShareLinks();
    if (!links[token]) return false;
    delete links[token];
    await saveShareLinks(links);
    return true;
  });
}

export async function getShareLink(token: string): Promise<ShareLink | null> {
  const links = await getAllShareLinks();
  return links[token] || null;
}

export function isExpired(link: ShareLink): boolean {
  if (!link.expiresAt) return false;
  return new Date(link.expiresAt).getTime() < Date.now();
}

export function verifyPin(link: ShareLink, pin?: string): boolean {
  if (!link.pinHash) return true; // 无密码
  if (!pin) return false;
  return hashPin(pin) === link.pinHash;
}

export async function shareTokenExists(token: string): Promise<boolean> {
  const link = await getShareLink(token);
  return !!link;
}

export async function getAllShares(): Promise<Array<{ token: string } & ShareLink>> {
  const links = await getAllShareLinks();
  return Object.entries(links).map(([token, link]) => ({ token, ...link }));
}

File: backend/lib/share.py
import asyncio
import json
import os
from datetime import datetime, timezone

from backend.config import SHARE_LINKS_FILE

_lock = asyncio.Lock()


async def _init_file() -> None:
    if not SHARE_LINKS_FILE.exists():
        await asyncio.to_thread(lambda: SHARE_LINKS_FILE.write_text("{}", "utf-8"))


async def _read_all() -> dict:
    await _init_file()
    raw = await asyncio.to_thread(lambda: SHARE_LINKS_FILE.read_text("utf-8"))
    return json.loads(raw or "{}")


async def _save_all(links: dict) -> None:
    data = json.dumps(links, ensure_ascii=False, indent=2)
    await asyncio.to_thread(lambda: SHARE_LINKS_FILE.write_text(data, "utf-8"))


async def create_share_link(item_path: str, link_type: str) -> str:
    async with _lock:
        links = await _read_all()
        while True:
            token = os.urandom(16).hex()
            if token not in links:
                break

        links[token] = {
            "path": item_path,
            "type": link_type,
            "createdAt": datetime.now(timezone.utc).isoformat(),
        }
        await _save_all(links)
        return token


async def delete_share_link(token: str) -> bool:
    async with _lock:
        links = await _read_all()
        if token not in links:
            return False
        del links[token]
        await _save_all(links)
        return True


async def get_share_link(token: str) -> dict | None:
    links = await _read_all()
    return links.get(token)


async def get_all_shares() -> list[dict]:
    links = await _read_all()
    return [{"token": token, **link} for token, link in links.items()]
